### PR TITLE
Fix(optimizer): only qualify coalesced USING columns if they exist in table schemas

### PR DESCRIPTION
**Pull request by georgesittas**

### CODE REVIEW

File: sqlglot/optimizer/qualify_columns.py [hunk 1]
```diff
@@ -776,6 +785,10 @@ def all_columns(self) -> t.Set[str]:
 
     def get_source_columns(self, name: str, only_visible: bool = False) -> t.Sequence[str]:
         """Resolve the source columns for a given source `name`."""
+        source_columns = self._get_source_columns_cache.get((name, only_visible))
```
Comment (tobymao) [inline sqlglot/optimizer/qualify_columns.py:788]: i like the pattern of

if ... not in ...:
    self._source_columns = ...
return self._source_column[(name, only_visible)]

File: sqlglot/optimizer/qualify_columns.py [hunk 2]
```diff
@@ -724,6 +732,7 @@ def __init__(self, scope: Scope, schema: Schema, infer_schema: bool = True):
         self._unambiguous_columns: t.Optional[t.Mapping[str, str]] = None
         self._all_columns: t.Optional[t.Set[str]] = None
         self._infer_schema = infer_schema
+        self._get_source_columns_cache: t.Dict[t.Tuple[str, bool], t.Sequence[str]] = {}
```
[the remark this answers is not in the document]
Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py:735]: We already have `_source_columns` above

[the remark this answers is not in the document]
Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py:735]: Perhaps we can consolidate the two, lemme take a quick look..

File: sqlglot/optimizer/qualify_columns.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/optimizer/qualify_columns.py:788]: yea

Comment (georgesittas) [reply sqlglot/optimizer/qualify_columns.py]: [Done](https://github.com/tobymao/sqlglot/pull/3740/commits/edcca349198e68b8c39980d09d30a3e60cd40c73)

**Review by tobymao — APPROVED** [review]

**Pull request merged**